pkg/sql/colexec/execgen/inline.go

```diff
@@ -0,0 +1,363 @@
+			reassignmentStmt := getFormalParamReassignments(decl, callExpr)
+			inlinedStatements.List = append(inlinedStatements.List, &dst.BlockStmt{
+				List: append([]dst.Stmt{reassignmentStmt}, body.List...),
+			})
+			// Assign mangled return values to the original assignment variables.
+			newAssignment := dst.Clone(n).(*dst.AssignStmt)
+			newAssignment.Rhs = make([]dst.Expr, len(retValNames))
+			for i := range retValNames {
+				newAssignment.Rhs[i] = dst.NewIdent(retValNames[i])
+			}
+			inlinedStatements.List = append(inlinedStatements.List, newAssignment)
+			cursor.Replace(inlinedStatements)
+
+		case *dst.ExprStmt:
+			// Search for raw function call:
+			// foo()
+			callExpr, ok := n.X.(*dst.CallExpr)
+			if !ok {
+				return true
+			}
+			decl := getTemplateFunc(templateFuncMap, callExpr)
+			if decl == nil {
+				return true
+			}
+
+			reassignments := getFormalParamReassignments(decl, callExpr)
+
+			// This case is simpler than the AssignStmt case. It's identical, except
+			// there is no mangled return value name block, nor re-assignment to
+			// the mangled returns after the inlined function.
+			funcBlock := &dst.BlockStmt{
+				List: []dst.Stmt{reassignments},
+			}
+			body := dst.Clone(decl.Body).(*dst.BlockStmt)
+
+			// Remove return values if there are any, since we're ignoring returns
+			// as a raw function call.
+			body = replaceReturnStatements(decl.Name.Name, body, nil)
+			// Add the inlined function body to the block.
+			funcBlock.List = append(funcBlock.List, body.List...)
+
+			cursor.Replace(funcBlock)
+		}
+		return true
+	}, nil)
+
+	b := bytes.Buffer{}
+	_ = decorator.Fprint(&b, f)
+	return b.String(), nil
+}
+
+// extractInlineFuncDecls searches the input file for functions that are
+// annotated with execgen:inline, extracts them into templateFuncMap, and
+// deletes them from the AST.
+func extractInlineFuncDecls(f *dst.File, templateFuncMap map[string]*dst.FuncDecl) dst.Node {
+	return dstutil.Apply(f, func(cursor *dstutil.Cursor) bool {
+		n := cursor.Node()
+		switch n := n.(type) {
+		case *dst.FuncDecl:
+			var mustInline bool
+			for _, dec := range n.Decorations().Start.All() {
+				if dec == "// execgen:inline" {
+					mustInline = true
+					break
+				}
+			}
+			if !mustInline {
+				// Nothing to do, but recurse further.
+				return true
+			}
+			for _, p := range n.Type.Params.List {
+				if len(p.Names) > 1 {
+					panic("can't currently deal with multiple names per type in decls")
+				}
+			}
+			// Store the function in a map.
+			templateFuncMap[n.Name.Name] = n
+			// Replace the function textually with a fake constant. We do this instead
+			// of completely deleting it to prevent "important comments" above the
+			// function to be deleted, such as template comments like {{end}}. This
+			// is kind of a quirk of the way the comments are parsed, but nonetheless
+			// this is an easy fix so we'll leave it for now.
+			cursor.Replace(&dst.GenDecl{
```

Comment:
Just to understand, this emits a "_ = `inlined_func`" where the function call was?

Reply:
Yeah. Added a comment.